content/blog/waddlesplash/2023-04-24_condition_variables.md

```diff
@@ -0,0 +1,416 @@
+
+Different compilers, operating systems, etc. expose such ISA-level atomic operations in different ways. On Haiku, we expose them through a set of functions beginning with `atomic_` which operate on `int32*`s (and there is also an equivalent set of functions for `int64*`s.) If we use one of them to make our earlier example thread-safe, it will look like this:
+```cpp
+void add() {
+	atomic_add(gSharedInteger, 1);
+}
+```
+(As a bonus, `atomic_add` also returns the value `gSharedInteger` had before `1` was added to it: so it is really a “fetch and add” operation, not just an “add” operation.)
+
+There are a variety of different atomic operations, and many ways to use them. But it is important to keep in mind that these are a ‘low-level’ primitive, and when using them, there are a lot of considerations that must be taken into account: when using multiple atomic operations in a row, interesting things can happen with memory ordering, for example.
+
+### Spinlocks
+
+Atomic operations mostly serve as a “building block” upon which more ergonomic synchronization primitives are built. Foremost among these, underlying all other thread synchronization primitives in the kernel, is the *spinlock*.
+
+```cpp
+spinlock gLock = B_SPINLOCK_INITIALIZER;
+int32 gSharedInteger = 0;
+
+void add() {
+	acquire_spinlock(&gLock);
+	gSharedInteger = gSharedInteger + 1;
+	release_spinlock(&gLock);
+}
+```
+
+As the name implies, spinlocks simply “spin” (run atomic operations in a loop) until they manage to acquire the lock. They can thus wind up wasting lots of CPU time if another core holds them for a long period of time. (Haiku’s implementation has mechanisms to detect this and trigger kernel panics when it happens.) They are mostly used in implementing other synchronization primitives, and in especially low-level parts of the kernel.
+
+Any thread which is acquiring or holding spinlocks is not allowed to be rescheduled, or even interrupted for any reason. Thus, spinlocks are always acquired only after the CPU’s “interrupts disabled” mode has been activated, and attempts to acquire them outside this mode will fault.
+
+### Semaphores
+
+The next kind of thread synchronization primitive, and in fact the only high-level one BeOS had in its kernel APIs, is called a *semaphore*, which tracks abstract “units” and lets threads wait for them to become available. For example:
+
+```cpp
+sem_id gSemaphore = create_sem(1, "test semaphore");
+int32 gSharedInteger = 0;
+
+void add() {
+	if (acquire_sem(gSemaphore) != B_OK)
+		return;
+	gSharedInteger = gSharedInteger + 1;
+	release_sem(gSemaphore);
+}
+```
+
+Here, I have created a semaphore which starts out having 1 “unit”. Within `add`, I try to acquire a “unit” from the semaphore; if there is not one available, the thread in question will become *blocked*, and will only be woken up and resume execution once the semaphore can be acquired.
```

Comment:
Maybe make it explicit that there is an upside and downside compared to spinlocks:

Upside: a blocked thread will release the CPU for other threads to run while it is blocked
Downside: this means there will be a context switch when the thread is blocked, and that is a relatively slow operation